lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1973,6 +1996,95 @@ class DecomposeAtenEinsumOp : public OpRewritePattern<AtenEinsumOp> {
 };
 } // namespace
 
+namespace {
+// Trilinear einstein sum, decomposed to:
+// (i1.unsqueeze(expand1) * i2.unsqueeze(expand2) * i3.unsqueeze(expand3))
+//    .sum(sumdim)
+
+class DecomposeAten_TrilinearOp : public OpRewritePattern<Aten_TrilinearOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(Aten_TrilinearOp op,
+                                PatternRewriter &rewriter) const override {
+
+    Location loc = op.getLoc();
+
+    Value input1 = op.getI1();
+    Value input2 = op.getI2();
+    Value input3 = op.getI3();
+
+    // Expansions
+    SmallVector<int64_t> expand1;
+    SmallVector<int64_t> expand2;
+    SmallVector<int64_t> expand3;
+    if (!matchPattern(op.getExpand1(), m_TorchListOfConstantInts(expand1))) {
+      return rewriter.notifyMatchFailure(op, "expand1 should be constant");
+    }
+    if (!matchPattern(op.getExpand2(), m_TorchListOfConstantInts(expand2))) {
+      return rewriter.notifyMatchFailure(op, "expand2 should be constant");
+    }
+    if (!matchPattern(op.getExpand3(), m_TorchListOfConstantInts(expand3))) {
+      return rewriter.notifyMatchFailure(op, "expand3 should be constant");
+    }
```

Comment:
What about the case where something lies in the triple intersecton of the expand sets? I thought we were going to handle that case.

Reply:
Yeah, that's the case where it inserts the dimension of zero. I was thinking of leaving it as-is since the actual einsum op doesn't zero out that dimension:

```
>>> a, b, c = [torch.rand(1, 3, 3, 3) for i in range(3)] # Simulate expand at dim=0 for all input tensors
>>> torch.einsum('abcd,abcd,abcd->abcd', a, b, c).shape
torch.Size([1, 3, 3, 3])
```

But, I could see how that'd be confusing downstream. Should I go back to using "Empty tensor or match failure" in that case?

Reply:
If you genuinely think there is a bug in the pytorch implementation, we could simply do something like 

```cpp
if (sharedExpand) {
   // pytorch issue filed: <link>
  // TODO: remove warning when issue gets resolved. 
   op->EmitWarning("aten::_trilinear implementation in this case is non-functional (returns an empty tensor). We will intentionally deviate from this behavior.");
```

Reply:
PyTorch issue filed [here](https://github.com/pytorch/pytorch/issues/138353), and emitWarning added for this case